src/Stratis.Bitcoin.Features.Consensus/StakeValidator.cs

```diff
@@ -143,26 +308,78 @@ private static uint256 ToUInt256(BigInteger input)
+            bool res = actualDepth < maxDepth;
+
+            this.logger.LogTrace("(-):{0}", res);
+            return res;
+        }
+
+        /// <summary>
+        /// Verifies transaction's signature.
+        /// </summary>
+        /// <param name="txFrom">UTXO that is spent in the transaction.</param>
+        /// <param name="txTo">Transaction.</param>
+        /// <param name="txToInN">Index of the transaction's input.</param>
+        private bool VerifySignature(UnspentOutputs txFrom, Transaction txTo, int txToInN, ScriptVerify flagScriptVerify)
+        {
+            this.logger.LogTrace("({0}:'{1}/{2}',{3}:{4},{5}:{6})", nameof(txFrom), txFrom.TransactionId, txFrom.Height, nameof(txToInN), txToInN, nameof(flagScriptVerify), flagScriptVerify);
+
+            TxIn input = txTo.Inputs[txToInN];
+
+            if (input.PrevOut.N >= txFrom._Outputs.Length)
+                return false;
+
+            if (input.PrevOut.Hash != txFrom.TransactionId)
+                return false;
+
+            TxOut output = txFrom._Outputs[input.PrevOut.N];
+
+            var txData = new PrecomputedTransactionData(txTo);
+            var checker = new TransactionChecker(txTo, txToInN, output.Value, txData);
+            var ctx = new ScriptEvaluationContext { ScriptVerify = flagScriptVerify };
+
+            bool res = ctx.VerifyScript(input.ScriptSig, output.ScriptPubKey, checker);
+            this.logger.LogTrace("(-):{0}", res);
+            return res;
+        }
+
+        /// <summary>
+        /// Stratis kernel protocol. Checks the stake kernel hash.
+        /// </summary>
+        /// <remarks>
+        /// Coinstake must meet hash target according to the protocol:
+        /// kernel (input 0) must meet the formula
+        /// <c>hash(stakeModifierV2 + stakingCoins.Time + prevout.Hash + prevout.N + transactionTime) &lt; target * weight</c>.
+        /// This ensures that the chance of getting a coinstake is proportional to the amount of coins one owns.
+        /// <para>
+        /// The reason this hash is chosen is the following:
+        /// <list type="number">
+        /// <item>stakeModifierV2: Scrambles computation to make it very difficult to precompute future proof-of-stake.</item>
+        /// <item>stakingCoins.Time: Time of the coinstake UTXO. Slightly scrambles computation.</item>
+        /// <item>prevout.Hash: Hash of stakingCoins UTXO, to reduce the chance of nodes generating coinstake at the same time.</item>
+        /// <item>prevout.N: Output number of stakingCoins UTXO, to reduce the chance of nodes generating coinstake at the same time.</item>
+        /// <item>transactionTime: Timestamp of the coinstake transaction.</item>
+        /// </list>
+        /// Block or transaction tx hash should not be used here as they can be generated in vast
+        /// quantities so as to generate blocks faster, degrading the system back into a proof-of-work situation.
+        /// </para>
+        /// </remarks>
```

Comment:
in general, those old comments are old, they are not necessary good comments, so i'd prefer rewriting them. just styling them like this is better than nothing, but could be much better
also you are lacking lots of `<see>`s here